File: 10-12-2023_Duplicate_Subtree_in_Binary_Tree.cpp
class Solution {
  public:
    /*This function returns true if the tree contains 
    a duplicate subtree of size 2 or more else returns false*/
    string structure(unordered_map<string, int> &mp, Node* root){
        if(root == NULL){
            return "";
        }
        string path = structure(mp, root -> left) + to_string(root -> data) + structure(mp, root -> right);
        mp[path]++;
        return path;
    }
    int dupSub(Node *root){
        unordered_map<string, int> mp;
        string result = structure(mp, root);
        for(auto it : mp){
            if(it.first.length() > 2 and it.second > 1){
                return 1;
            }
        }
        return 0;
    }
};

File: 10-12-2023_Duplicate_Subtree_in_Binary_Tree.py
class Solution:
    def __init__(self):
        self.mp = {}
    def structure(self, root):
        if root is None:
            return ""
        path = self.structure(root.left) + str(root.data) + self.structure(root.right)
        if path in self.mp:
            self.mp[path] += 1
        else:
            self.mp[path] = 1
        return path
    def dupSub(self, root):
        result = self.structure(root)
        for key, value in self.mp.items():
            if len(key) > 2 and value > 1:
                return 1
        return 0
